### PR TITLE
[SPARK-34199][SQL] Block `count(table.*)` to follow ANSI standard and other SQL engines

### PR DESCRIPTION
**Pull request by linhongliu-db**
### What changes were proposed in this pull request?
In spark, the `count(table.*)` may cause very weird result, for example:
```
select count(*) from (select 1 as a, null as b) t;
output: 1
select count(t.*) from (select 1 as a, null as b) t;
output: 0
```
 This is because spark expands `t.*` while converts `*` to count(1), this will confuse
users. After checking the ANSI standard, `count(*)` should always be `count(1)` while `count(t.*)`
is not allowed. What's more, this is also not allowed by common databases, e.g. MySQL, Oracle.

So, this PR proposes to block the ambiguous behavior and print a clear error message for users.

### Why are the changes needed?
to avoid ambiguous behavior and follow ANSI standard and other SQL engines

### Does this PR introduce _any_ user-facing change?
Yes, `count(table.*)` behavior will be blocked and output an error message.

### How was this patch tested?
newly added and existing tests

### CODE REVIEW

**Comment by linhongliu-db** [comment]
cc @cloud-fan

**Comment by linhongliu-db** [comment]
cc @cloud-fan, this PR is ready to review

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -89,6 +89,8 @@ license: |
     * `ALTER TABLE .. ADD PARTITION` throws `PartitionsAlreadyExistException` if new partition exists already
     * `ALTER TABLE .. DROP PARTITION` throws `NoSuchPartitionsException` for not existing partitions
 
+  - In Spark 3.1, usage of `count(table.*)` is blocked to avoid producing ambiguous results. Because `count(*)` and `count(table.*)` will output differently if there is any null values. To restore the behavior before spark 3.1, you can set `spark.sql.legacy.allowSingleTableStarInCount` to `true`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md:92]: We are in 3.2 now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -250,6 +250,12 @@ object QueryCompilationErrors {
     new AnalysisException(s"Invalid usage of '*' in $prettyName")
   }
 
+  def tableStarInCountNotAllowedError(targetString: String): Throwable = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:253]: `singleTableStar...`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -250,6 +250,12 @@ object QueryCompilationErrors {
...
+    new AnalysisException(s"It's not allowed to qualify $targetString.* " +
+      "using table name inside a count function. " +
+      "Please use '*' only or expand the table manually, e.g. court(*), count(col1, col2)")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:256]: how about `count($targetString.*) is now allowed. Please use ...`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:256]: `expand the table` -> `expand the columns`

File: sql/core/src/test/resources/sql-tests/inputs/count.sql [hunk 4]
```diff
@@ -19,11 +19,11 @@ SELECT
 FROM testData;
 
 -- count with multiple expressions
-SELECT count(a, b), count(b, a), count(testData.*) FROM testData;
+SELECT count(a, b), count(b, a), count(testData.a, testData.b) FROM testData;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/count.sql:22]: how about `count(testData.*, testData.*)`

File: sql/core/src/test/resources/sql-tests/inputs/count.sql [hunk 5]
```diff
@@ -19,11 +19,11 @@ SELECT
...
 
 -- distinct count with multiple expressions
 SELECT
-  count(DISTINCT a, b), count(DISTINCT b, a), count(DISTINCT *), count(DISTINCT testData.*)
+  count(DISTINCT a, b), count(DISTINCT b, a), count(DISTINCT *), count(DISTINCT testData.a, testData.b)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/count.sql:26]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql [hunk 6]
```diff
@@ -743,20 +743,20 @@ select * from a left join b on i = x and i = y and x = i;
 --
 -- test NULL behavior of whole-row Vars, per bug #5025
 --
-select t1.q2, count(t2.*)
+select t1.q2, count(t2.q1, t2.q2)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: `count(t2.*)` means `count(1)` in pgsql?

**Comment by cloud-fan** [comment]
makes sense to me, cc @viirya @maropu

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 6 above)
Comment (linhongliu-db) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: Yes, in pgsql
```
select * from (select 1 as a, null as b) as t1 join (select 1 as a, 'c' as b) as t2 on t1.a = t2.a
output:

a  | b      | a  | b
-- | --     | -- | --
1  | (null) | 1  | c

select count(t1.*) from (select 1 as a, null as b) as t1 join (select 1 as a, 'c' as b) as t2 on t1.a = t2.a
output:

count
--
1
``

Comment (linhongliu-db) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: in spark, the second query will output 0

File: sql/core/src/test/resources/sql-tests/inputs/count.sql (hunk 4 above)
Comment (linhongliu-db) [reply sql/core/src/test/resources/sql-tests/inputs/count.sql]: sure, I changed this in both count.sql and udf-count.sql.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: since it's pgsql test, shall we use `count(1)` to match pgsql test intention?

File: docs/sql-migration-guide.md [hunk 7]
```diff
@@ -49,6 +49,8 @@ license: |
     * and the method `spark.catalog.refreshTable`
   In Spark 3.1 and earlier, table refreshing leaves dependents uncached.
 
+  - In Spark 3.2, usage of `count(table.*)` is blocked to avoid producing ambiguous results. Because `count(*)` and `count(table.*)` will output differently if there is any null values. To restore the behavior before spark 3.2, you can set `spark.sql.legacy.allowSingleTableStarInCount` to `true`.
```
Comment (viirya) [inline docs/sql-migration-guide.md:52]: "before spark 3.2" -> "before Spark 3.2"

Comment (viirya) [inline docs/sql-migration-guide.md:52]: "usage" -> "the usage"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -1763,6 +1763,21 @@ class Analyzer(override val catalogManager: CatalogManager)
     def expandStarExpression(expr: Expression, child: LogicalPlan): Expression = {
       expr.transformUp {
         case f1: UnresolvedFunction if containsStar(f1.arguments) =>
+          // SPECIAL CASE: We want to block count(table.*) because in spark, count(table.*) will
+          // be expanded while count(*) will be converted to count(1). They will produce different
+          // results and confuse users if there is any null values. For count(t1.*, t2.*), it is
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1768]: Why `count(t1.*, t2.*)` is well-defined but `count(table.*)` is not? Isn't `count(t1.*, t2.*)` also expended like `count(table.*)` too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -1559,6 +1559,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val ALLOW_SINGLE_TABLE_STAR_IN_COUNT =
+    buildConf("spark.sql.legacy.allowSingleTableStarInCount")
+      .internal()
+      .doc("When true, the SQL function 'count' is allowed to take single 'table.*' as parameter")
+      .version("3.1.1")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1566]: `version("3.2")`

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 10]
```diff
@@ -155,6 +155,21 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
     checkAnswer(testData.as("testData").select($"testData.*"), testData.collect().toSeq)
   }
 
+  test("star can be qualified by table name inside a non-count function") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:158]: Please add JIRA number in the test name: "SPARK-34199: star can ..."

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 11]
```diff
@@ -155,6 +155,21 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
...
+    checkAnswer(
+      testData.as("testData").selectExpr("hash(testData.*)"),
+      testData.as("testData").selectExpr("hash(testData.key, testData.value)")
+    )
+  }
+
+  test("star cannot be qualified by table name inside a count function") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:165]: ditto.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (linhongliu-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1768]: because expand `count(t1.*, t2.*)` is not ambiguous, I think no one will argue that `count(t1.*, t2.*)` should equal `count(1)`. Also this usage is not supported by the common database (MySQL, oracle, pgsql), so we are not in conflict with others.
but `count(table.*)` is not same. spark SQL expands the columns, pgsql converts it to `count(1)`, MySQL, Oracle (as well as ANSI) thinks this should be disallowed.
So I think block `count(table.*)` can avoid ambiguous while keeping `count(t1.*, t2.*)` can lead less side effects.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 6 above)
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: +1 for the @cloud-fan advice. Also, could you leave some comment there about why we change the original test query?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 12]
```diff
@@ -1559,6 +1559,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val ALLOW_SINGLE_TABLE_STAR_IN_COUNT =
+    buildConf("spark.sql.legacy.allowSingleTableStarInCount")
```
[the remark this answers is not in the document]
Comment (linhongliu-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1563]: allowStarWithSingleTableIdentifierInCount? Should we make this more clear?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1768]: How about the case `select count(t.*, t.*) from values (1, null) t(a, b)`? Btw, If the other databases don't support the syntax, `count(t1.*, t2.*)`, we still need to keep supporting it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1563]: sgtm

**Comment by maropu** [comment]
Disallowing the syntax looks fine.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (linhongliu-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1768]: Thanks @maropu  for reviewing.

`select count(t.*, t.*) from values (1, null) t(a, b)` will output 0.

I'm fine with blocking `count(t1.*, t2.*)` as well but since in spark, we are allowing other similar cases that other databases don't support (and not follow ANSI), it's not harmful to keep `count(t1.*, t2.*)` as one more case. After all, introducing unnecessary behavior change (blocking `count(t1.*, t2.*)`) doesn't benefit users.
The similar usages:
count(col_a, col_b) - count multiple columns is not supported by pgsql, oracel. MySQL only support with distinct
count(struct_col.*) - expand columns in struct data type is not supported by the mentioned databases.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1768]: Let's minimize the change for now. We can open a new PR to discuss if we should forbid other cases.

**Review by viirya — COMMENTED** [review]
Looks okay to me.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 6 above)
Comment (linhongliu-db) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: @cloud-fan , @maropu , I did more tests on pgsql, seems it is much more wired than I thought.
TLDR: in pgsql, `count(t2.*)` means how many MATCHING rows from t2 table. and spark doesn't have such a semantic. Here, if we want to keep the output unchanged. it should be `count(t2.q1, t2.q2)`

Here is the example:
```
CREATE TABLE INT8_TBL AS SELECT * FROM (VALUES (1, 2), (2, 1), (3, NULL), (NULL, NULL)) AS v(q1, q2);
select * from int8_tbl as t1 left join int8_tbl as t2 on t1.q1 = t2.q2;

t1.q1  | t1.q2  | t2.q1  | t2.q2
--     | --     | --     | --
1      | 2      | 2      | 1
2      | 1      | 1      | 2
3      | (null) | (null) | (null)
(null) | (null) | (null) | (null)

select count(t1.*), count(t2.*) from int8_tbl as t1 left join int8_tbl as t2 on t1.q1 = t2.q2;

t1.count | t2.count
--       | --
4        | 2

```

File: docs/sql-migration-guide.md [hunk 13]
```diff
@@ -49,6 +49,8 @@ license: |
     * and the method `spark.catalog.refreshTable`
   In Spark 3.1 and earlier, table refreshing leaves dependents uncached.
 
+  - In Spark 3.2, the usage of `count(table.*)` is blocked to avoid producing ambiguous results. Because `count(*)` and `count(table.*)` will output differently if there is any null values. To restore the behavior before Spark 3.2, you can set `spark.sql.legacy.allowStarWithSingleTableIdentifierInCount` to `true`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md:52]: `table` is a keyword, so `count(table.*)` is not parse-able under ansi mode. How about `count(tblName.*)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 14]
```diff
@@ -1763,6 +1763,21 @@ class Analyzer(override val catalogManager: CatalogManager)
     def expandStarExpression(expr: Expression, child: LogicalPlan): Expression = {
       expr.transformUp {
         case f1: UnresolvedFunction if containsStar(f1.arguments) =>
+          // SPECIAL CASE: We want to block count(table.*) because in spark, count(table.*) will
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1766]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 15]
```diff
@@ -1559,6 +1559,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val ALLOW_STAR_WITH_SINGLE_TABLE_IDENTIFIER_IN_COUNT =
+    buildConf("spark.sql.legacy.allowStarWithSingleTableIdentifierInCount")
+      .internal()
+      .doc("When true, the SQL function 'count' is allowed to take single 'table.*' as parameter")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1565]: ditto

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 6 above)
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: nit: could you leave some comments here about why we updated this test?

Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:746]: nit: Could you leave some comments here about why we updated the tests?

**Review by maropu — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 6 above)
Comment (linhongliu-db) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql:751]: fixed

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
